File: kSortedList.java
// Time Complexity : O(N*L) where N is the number of heads and L is the length of the list
// Space Complexity : O(N) N = number of nodes
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/*
	1. Add all the heads in the minHeap
	2. Remove the smallest one and add to our final result LL
	3. Add the next element of the recently removed node.
	4. Return our custom LL containing the elements in increasing order
*/


// Your code here along with comments explaining your approach
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a,b) -> a.val - b.val);
        ListNode temp = new ListNode(-1);
        ListNode curr = temp;
        
        for(ListNode list: lists) {
            if(list!=null)
                minHeap.add(list);
        }
        
        while(!minHeap.isEmpty()) {
            ListNode min = minHeap.poll();
            curr.next = min;
            curr = curr.next;
            if(min.next!=null) minHeap.add(min.next);
        }
        return temp.next;
        
    }
}

File: KSortedList.py
# Time Complexity : O(nk logk) Where k is the largest Number
# Space Complexity : O(k)

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        ListNode.__lt__ = (lambda a,b : a.val < b.val)

        heap = []
        dummy = ListNode(None, None)
        result = dummy

        for head in lists:
            if head:
                heapq.heappush(heap, head)

        while heap:
            current_smallest = heapq.heappop(heap)
            if current_smallest.next:
                heapq.heappush(heap, current_smallest.next)
            dummy.next = current_smallest
            dummy = dummy.next

        return result.next